### PR TITLE
Potential field - potential range and ocillations

### PR DESCRIPTION
**Pull request by adi-vardi-enway**
2 changes:
1.  Potential field cover both start and goal - previously potential field was only defined at a certain area around any obstacles. So if goal/start are outside the area, the potential field is undefined, leading to the planner not converging. This fix extends the potential filed to include the start and goal if they are outside the obstacles area.
2. Add simple oscillations detection with horizon of the last 3 previous poses.

### CODE REVIEW

**Comment by AtsushiSakai** [comment]
@adi-vardi-1 Thank you for great PR!!. I have some review comments.

File: PathPlanning/PotentialFieldPlanning/potential_field_planning.py [hunk 1]
```diff
@@ -106,14 +106,17 @@ def potential_field_planning(sx, sy, gx, gy, ox, oy, reso, rr):
 
     rx, ry = [sx], [sy]
     motion = get_motion_model()
+    previous_id = [(None, None)] * 3
```
Comment (AtsushiSakai) [inline PathPlanning/PotentialFieldPlanning/potential_field_planning.py:109]: Maybe the variable name should be "xy_history" or something like that. And number 3 should be global parameters like "AREA_WIDTH".

File: PathPlanning/PotentialFieldPlanning/potential_field_planning.py [hunk 2]
```diff
@@ -20,11 +20,11 @@
 show_animation = True
 
 
-def calc_potential_field(gx, gy, ox, oy, reso, rr):
-    minx = min(ox) - AREA_WIDTH / 2.0
-    miny = min(oy) - AREA_WIDTH / 2.0
-    maxx = max(ox) + AREA_WIDTH / 2.0
-    maxy = max(oy) + AREA_WIDTH / 2.0
+def calc_potential_field(gx, gy, ox, oy, reso, rr, sx, sy):
```
Comment (AtsushiSakai) [inline PathPlanning/PotentialFieldPlanning/potential_field_planning.py:23]: good catch. you are right. thank you.

File: PathPlanning/PotentialFieldPlanning/potential_field_planning.py [hunk 3]
```diff
@@ -128,6 +131,19 @@ def potential_field_planning(sx, sy, gx, gy, ox, oy, reso, rr):
         rx.append(xp)
         ry.append(yp)
 
+        if ((None, None) not in previous_id and
```
Comment (AtsushiSakai) [inline PathPlanning/PotentialFieldPlanning/potential_field_planning.py:134]: I think this if evaluation should be a boolean function like "is_oscilication_detected" for readability.

File: PathPlanning/PotentialFieldPlanning/potential_field_planning.py [hunk 4]
```diff
@@ -128,6 +131,19 @@ def potential_field_planning(sx, sy, gx, gy, ox, oy, reso, rr):
...
+                (previous_id[0] == previous_id[1]
+                 or previous_id[1] == previous_id[2]
+                 or previous_id[0] == previous_id[2])):
+            print("Oscillation detected!!!")
+            print(previous_id)
+            break
+
+        # roll previous
+        previous_id[0] = previous_id[1]
```
Comment (AtsushiSakai) [inline PathPlanning/PotentialFieldPlanning/potential_field_planning.py:143]: This code can be used for only when horizon is 3. So, I think "deque" should be used here. https://docs.python.org/3/library/collections.html#collections.deque

Comment (adi-vardi-enway) [reply PathPlanning/PotentialFieldPlanning/potential_field_planning.py]: done

Comment (adi-vardi-1) [reply PathPlanning/PotentialFieldPlanning/potential_field_planning.py]: done

File: PathPlanning/PotentialFieldPlanning/potential_field_planning.py (hunk 3 above)
Comment (adi-vardi-1) [reply PathPlanning/PotentialFieldPlanning/potential_field_planning.py]: good point, done

File: PathPlanning/PotentialFieldPlanning/potential_field_planning.py (hunk 1 above)
Comment (adi-vardi-1) [reply PathPlanning/PotentialFieldPlanning/potential_field_planning.py]: added

**Comment by adi-vardi-enway** [comment]
great :)

**Pull request merged**